### PR TITLE
libobs-d3d11, libobs-opengl: Don't deallocate foreign memory

### PR DESCRIPTION
**Pull request by Xaymar-Streamlabs**
Changes the deallocation behaviour of either data to be manual in order to allow the calling application to use predetermined points in memory without encountering unexpected crashes.

Fixes: https://obsproject.com/mantis/view.php?id=907

### CODE REVIEW

**Comment by ONSBalder** [comment]
Can one of the admins verify this patch?

**Comment by Xaymar** [comment]
@jp9000 Is there any change of this getting merged in the near future?
I have plugins depending on correct graphics subsystem behaviour and this is currently crashing every single one of them. The only way to work around this issue is to duplicate the memory every time I want to do any changes to a vertex buffer and hope that the graphics subsystem works as expected.

**Comment by jp9000** [comment]
What specific code can you point me to that's causing you trouble with this?  You're supposed to use memory allocated with bmalloc on the values in the gs_vb_data structure.

**Comment by Xaymar** [comment]
My [obs-stream-effects](https://github.com/Xaymar/obs-stream-effects/blob/master/source/gs-vertexbuffer.cpp) plugin currently works around the issue by memory patching the gs driver (D3D11 or OpenGL) to allow managed vertex data. The current behaviour does not allow for memory to be managed by anyone but the graphics subsystem, but the graphics subsystem is incredibly inconsistent with that. This PR fixes this inconsistency.

Another example for what would be really difficult to do is model loading. With the current approach I'd have to first load the model from disk, then copy the memory over into what OBS wants. If I also have a vertex-animated model, this copy has to be redone every frame, adding huge overhead. This also applies to the index buffers, which have the same issue.

Simply put, the gs subsystem is incorrectly deleting memory that it didn't create in the first place. No matter what, functions should not delete memory that was not created by a previous function of similar kind. For example, if gs_vertexbuffer_create did not create the memory, then gs_vertexbuffer_destroy should not destroy it either.

**Comment by jp9000** [comment]
Okay, I'm adding a flag to the parameters to make it so it optionally doesn't take ownership if requested -- instead it will make a copy and leave the gs_vbdata_t structure passed as-is, in case it was filled out with C++ vectors or something.  I admit that the "taking ownership" thing is unusual for an API, I probably should have made it copy the vertex data by default, but at the same time I don't want to break API.  However, just as a note, it copies, and to get its internal vertex data you will still have to use gs_vertexbuffer_get_data to get pointers to the actual buffers that will get flushed -- you won't be able to use the data of your std::vector to flush after initialization.

It would be like this regardless even had I not programmed it to take ownership of data.  I'm not going to design the API in such a way to where it will make the vertex buffer internally store pointers to foreign memory that it has to trust and flush memory that is foreign to itself without any sort of guarantee that the data is going to continue to exist -- so the options are either:  1.) take ownership, or 2.) copy the data and take ownership of the copy.  The idea that the vertex buffer should store pointers to foreign memory and hope that the memory is going to continue to exist is much worse than taking ownership of that memory by default like I have it programmed.  I do understand that there might be a use case to initialize the data more easily with foreign data such as C++ vectors, but the vertex buffer should never just store foreign pointers that it has no control over and just hope those pointers continue to exist for the vertex buffer's duration.  That would be worse API design, you have to understand that.

As for the use case of vertex animation, recreating a vertex buffer or index buffer every frame is not the correct approach, if you're doing it you need to re-examine your design.  If you need to increase vertex or index count dynamically, then set a specific maximum capacity for your vertex/index buffers, create vertex/index buffers with buffers set to that maximum capacity, and then perform a draw call specifying only the number of vertices/indices you want to use within the buffers.  You should never be recreating resources each frame if you can avoid it because it will negatively impact pipeline performance.

I will be adding the flag to copy the data along with the documentation branch.

Although I do admit that API has flaws, I do not want to change the API either -- I am very unlikely to merge patches that change API, just for future reference.  Additions to things like flags, or alternate versions of functions are fine, but API breakage is not fine, even if the API does have flaws.  I will not accept API design that is objectively worse either.

**Comment by Xaymar** [comment]
While this does fix the deallocation crash, it does not fix the inherent performance drop that comes with the proposed change. I will now be forced to copy this data each frame, which is far worse than just allowed external/foreign memory to be used as is.

Speaking of foreign memory, you do not "hope" that the memory continues to exist, you outright make the API user guarantee this. An API user that does not guarantee this is outright using the API wrong and is as such responsible for the crashes that happen due to this. The same could be said about this PR and obs-stream-effects, but at that point there was very little documentation or comments in the code that actually said what does what and how.

It is not "worse API design" to rely on foreign memory, it is actually done very commonly to drastically improve performance in those cases - in cases where this is not possible, the memory that is allocated by the function is versatile enough to store any kind of data you want to use. Most of the APIs you use actually do exactly this, especially when working with Hardware. 

As for vertex animations, you seem to have completely misunderstood what I said. I do not recreating any buffers each frame (except for texture caches if the texture changes) and updates are actually happening on the vertex information (position, normal, color, ...). The way the current and proposed API is going to work will add a second copy overhead from the third party provided buffer to whatever OBS gives me. While this is a fairly outdated animation technology, it does allow for things like morphing, which bone based animations do not.

Additionally you are being a bit condescending, if not straight up antagonizing, with your reply. You state "worse API design" as an objective fact, while it really is just an opinion/argument based on what you want the API to actually do. I can accept that you make a decision based on your opinions, I do not have to agree with these decisions or even "have to understand" it. Please separate your objective facts from opinions in the future, or at least try to make them sound less condescending.

**Comment by jp9000** [comment]
I think the proper solution to the situation that would solve this situation is
to really to just implement standard map/unmap functions for vertex buffers to
allow direct mapping/flushing of the data without having to rely on the
internal data, though the problem with that is that the vertex buffer needs a
copy of the vertex buffer data in case the driver crashes, that's the only
thing that really holds the whole thing up.  I suppose I could make it so that
it just uses old data for crashes if you really want to use a map/unmap instead
of flush.

Perhaps I did misunderstand or misread your vertex animation situation -- if so
I apologize for needlessly explaining that.  I can understand how that may have
seen condescending and probably didn't help.

As for the design debate, I will take back what I originally said about one
being worse over another:  both your design and my original design are at least
equally bad, because they both have crash cases -- that is what is truly
objective.  In my case, I will acknowledge that it was: "don't destroy the data
after passing ownership, otherwise there may be a crash".  However, with your
case, it's at least equally as bad because: "don't destroy/reallocate the data
until the vertex buffer is destroyed, otherwise there may be a crash".  These
are both bad -- to say otherwise is to have cognitive dissonance.  Neither is
normal.  It probably wouldn't be wise to go down the rabbit hole of arguing
about whether one is better or worse.

As for me being condescending -- let's not forget the "Workaround API stupidity
in obs-studio" hack as linked above, before I had even made the reply.  I think
we are both equally butting heads here.  Yes, I do get upset as well.  It's
normal for people to get frustrated with one another.  Programs, like
everything else in life, are made by people, almost always with flaws of some
sort.  However, make no mistake -- I do not consider myself "better than you"
or any other human being.  The fact that I made the "pass ownership" thing in
the first place with no other options is a clear sign that I am just as human
as you are, just as you made an equally bad solution.  I do not look down upon
you as some sort of peon just because I hold the executive power.  I
acknowledge your potential and skill, at least when you put ego and emotion
aside (as I hope you acknowledge I am doing here now).  You know I've told you
these things in the past.  If I were in your situation and I wanted to use an
API, but the API didn't allow for me to do the thing in the way that I
wanted/expected, I would probably be annoyed too.  If you had caught me when I
was designing it, perhaps I would have taken a different approach and realized
this sooner.

It is difficult for me to compromise much here because of a few factors, the
primary one being that I'm overworked at the moment and have higher priorities.

I want to implement a map/unmap for the actual "proper" solution to allow you
to use vertex data with vectors, but I'm not sure if I can do it right now.

Unfortunately, there are three options:

1.) Wait for me to eventually get around to making map/unmap functions,
2.) Create the map/unmap functions in libobs yourself without having to rely on
    me to do it for you (I don't think you'd enjoy that all that much), or
3.) Learn to use vertex buffers as they're currently designed:

    - Call gs_vbdata_create
    - Populate members of that structure according to what you need with
      bmalloc-allocated arrays
    - Pass that to gs_vertexbuffer_create
    - Use gs_vertexbuffer_get_data when you need the data, and modify it
      accordingly -- without using vectors.

I don't believe that last one is too hard.  That's how it's done throughout the
entirety of the rest of the program, even in my own C++ parts.  I realize it's
not what you want to hear, I realize I haven't been able to give you much in
return right now.  I realize you want to create your vertex buffer wrapper the
way you want to create it.

I'm not going to accept the changes proposed in the pull request because they
are no better than the way things currently are.  Allowing it would set a bad
precedence towards other contributors, saying "I allow people to code in hacks
for convenience."

**Comment by jp9000** [comment]
Alright, I figured out a simple and effective way to solve the issue:  https://github.com/jp9000/obs-studio/commit/a46f0b4 and https://github.com/jp9000/obs-studio/commit/c111fa6

The direct flush functions will allow you to use your own data and directly flush to the vertex/index buffers without having to go through their internal data.  These are easier to use than map/unmap functions as well, so that should be an extra boon for you.  The internal data will still optionally be present in your use case, but will not be used for anything except for rebuilding the vertex/index buffers in case the graphics driver crashes.  This way, you can use your own vertex/index buffer data buffers without having to rely on the internal data at all, and without having to perform extra unnecessary copies.  Use the GS_DUP_BUFFER flag to create your index/vertex buffers with your own buffer data, then the gs_vertexbuffer_flush_direct and gs_indexbuffer_flush_direct functions to directly flush using your own buffers.

Hopefully this is acceptable for you -- it satisfies both of our conditions.  I didn't want additional technical debt, and you didn't want to have to rely on the internal data.

The admittedly flawed design of "passing ownership" was a hand-me-down from my game engine.  I do like that design as an option mostly as a convenience, but I definitely don't think it should have been the default and especially the only way to use vertex/index buffers.  This PR was a result of that technical debt I myself introduced from that design, so this was ultimately my own doing, I fully admit.

Just in the future, two very important things I want out of code changes are:  that API isn't modified, and that technical debt does not increase (or is fixed rather than worked around).  I tend be pretty persistent when it comes to those two things.

**Comment by Xaymar** [comment]
The proposed fix works for me, I'll adjust my code once 21.x hits public.
